app/modules/mqtt.c

```diff
@@ -355,19 +482,81 @@ static void mqtt_socket_received(void *arg, char *pdata, unsigned short len)
       break;
 
     case MQTT_DATA:
-      mud->mqtt_state.message_length_read = length;
-      mud->mqtt_state.message_length = mqtt_get_total_length(in_buffer, mud->mqtt_state.message_length_read);
+      message_length = mqtt_get_total_length(in_buffer, in_buffer_length);
       msg_type = mqtt_get_type(in_buffer);
       msg_qos = mqtt_get_qos(in_buffer);
-      msg_id = mqtt_get_id(in_buffer, mud->mqtt_state.message_length);
+      msg_id = mqtt_get_id(in_buffer, in_buffer_length);
+
+      NODE_DBG("MQTT_DATA: msg length: %u, buffer length: %u\r\n",
+           message_length,
+           in_buffer_length);
+
+      if(message_length > mud->connect_info.max_message_length) {
+        // The pending message length is larger than we was configured to allow
+        NODE_DBG("MQTT: msg too long: total=%u, deliver=%u\r\n", message_length, in_buffer_length);
+        if (msg_type == MQTT_MSG_TYPE_PUBLISH) {
+          // In practice we should never get any other types..
```

Comment:
Even for QoS 2 messages?

Reply:
We could ignore the max_message_length for qos 2 if we want to, but question that follows: at what size should we crash instead of handling it gracefully?

Reply:
I think we should treat QoS 2 just like we treat QoS 1: delivery to "overflow" counts as delivery, even if we had to truncate the message.

Reply:
Yeah, not ACKing is going to be unfortunate.  It should be sufficient to just buffer the partial header and recombine it with the rest of the partial header that comes in a later packet, yes?

Reply:
Yes, as far as I know we can ACK as soon as we have the fixed+variable headers (topic + packet ID).
I guess that it should be done after we have "received" (skipped) the full message, not when we have the header(s), so we need to keep it until finished "receiving" the message.

Reply:
That sounds right.  I suspect some brokers would be... surprised if they got the ACK back before they were done sending the message!

Reply:
Pushed a new commit which ACKs overflowed messages, but doing this as soon as packet ID is known (if needed going into "short-buffer" mode first).
It does not wait for the final TCP segment to arrive, to keep things simple. Avoids having to keep more buffers & refs & states etc, and in most cases the sending broker will most likely have sent all of the PUBLISH to the OS socket anyway, being ready for an ACK.